File: python/src/agent_framework/rag/stores/in_memory.py
"""In-memory vector store."""

import math
from agent_framework.rag.types import SearchResult, VectorDocument


def _cosine_similarity(a: list[float], b: list[float]) -> float:
    if len(a) != len(b):
        return 0.0
    dot = sum(x * y for x, y in zip(a, b))
    norm_a = math.sqrt(sum(x * x for x in a))
    norm_b = math.sqrt(sum(x * x for x in b))
    denom = norm_a * norm_b
    return dot / denom if denom else 0.0


class InMemoryVectorStore:
    """In-memory vector store with cosine similarity."""

    def __init__(self) -> None:
        self._docs: list[VectorDocument] = []

    async def add(self, documents: list[VectorDocument]) -> None:
        self._docs.extend(documents)

    async def search(
        self,
        embedding: list[float],
        top_k: int,
        filter_: dict | None = None,
    ) -> list[SearchResult]:
        scored = [(doc, _cosine_similarity(doc.embedding, embedding)) for doc in self._docs]
        scored.sort(key=lambda x: x[1], reverse=True)
        return [SearchResult(doc, score) for doc, score in scored[:top_k]]

File: packages/rag/src/stores/in-memory.ts
/**
 * In-memory vector store: simple cosine similarity search.
 * For production, use pgvector, Chroma, etc.
 */
import type { VectorDocument, SearchResult } from "../types.js";
import type { VectorStore } from "../vector-store.js";

function cosineSimilarity(a: number[], b: number[]): number {
  if (a.length !== b.length) return 0;
  let dot = 0;
  let normA = 0;
  let normB = 0;
  for (let i = 0; i < a.length; i++) {
    dot += a[i] * b[i];
    normA += a[i] * a[i];
    normB += b[i] * b[i];
  }
  const denom = Math.sqrt(normA) * Math.sqrt(normB);
  return denom === 0 ? 0 : dot / denom;
}

/**
 * In-memory vector store (development / small datasets).
 */
export class InMemoryVectorStore implements VectorStore {
  private documents: VectorDocument[] = [];

  async add(documents: VectorDocument[]): Promise<void> {
    this.documents.push(...documents);
  }

  async search(
    embedding: number[],
    topK: number,
    _filter?: Record<string, unknown>
  ): Promise<SearchResult[]> {
    const scored = this.documents.map((doc) => ({
      document: doc,
      score: cosineSimilarity(doc.embedding, embedding),
    }));
    scored.sort((a, b) => b.score - a.score);
    return scored.slice(0, topK).map(({ document, score }) => ({
      document: {
        id: document.id,
        content: document.content,
        metadata: document.metadata,
        source: document.source,
        chunkIndex: document.chunkIndex,
        startChar: document.startChar,
        endChar: document.endChar,
      },
      score,
    }));
  }

  async delete(ids: string[]): Promise<void> {
    const set = new Set(ids);
    this.documents = this.documents.filter((d) => !set.has(d.id));
  }
}
